cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -85,67 +87,114 @@ message Result {
 
-#### Cons
+##### Cons
 
 * Reliance on gRPC call in the hot-path
+* Does not cache the data for endpoints that never had a DNS query.
 * In an event where SDP restarts when agent's gRPC service is unavailable, all cached state is lost and SDP cannot translate IP to endpoint ID or identity.
 
-### Option 2: Read from bpf maps
+#### Option 1b: Listen to ipcache updates via grpc
 
-#### Pros
+Similar to envoy, SDP can listen to ipcache updates and maintain a local cache of IP to endpoint ID mappings via grpc. This changes the way DNS proxy currently behaves i.e pull based model to push based model.
+
+##### Pros
+
+* Simpler to maintain and stay upto date with any new ip<>identity changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+##### Cons
+
+* Need a method for reconciling the cache in case of SDP restarts.
+* SDP will be aware of all the endpoints data, which might not be needed for DNS proxy.
+
+#### Option 2: Read from bpf maps [Recommended]
+
+We will need to read from couple of bpf maps to get the endpoint ID and identity for a given IP address. For ip to endpoint ID, we can read from `cilium_lxc` and for ip to identity, we can read from `cilium_ipcache`.
+
+##### Pros
 
 * Reduced resource utilization in agent since agent doesn't need to process rpc calls in hotpath
 * IP to endpoint ID or identity translation does not rely on agent availability.
+* DNS can be resolved even when agent is down.
 
-#### Cons
+##### Cons
 
 * Low level interactions from components external to cilium agent.
 
+### Q: Why did we choose to read from bpf maps ?
+
+The reason we chose to read from bpf maps is because the DNS proxy should be able to resolve the DNS queries even when agent is down. The bpf maps are the source of truth for cilium agent and are updated in real time. So, reading from bpf maps ensures that the DNS proxy is always in sync with the cilium agent.
+
+### Flag to disable built-in proxy
+
+Adding a flag(`--disable-builtin-dns-proxy`) to cilium agent to disable the built in dns proxy allows standlone dns proxy to run its independent lifecycle.
+
+#### Pros
+
+* Allows cilium agent to be upgraded without breaking the DNS traffic.
+* Provides a freedom to allow any delegated dns proxy to be plugged in.
+* Reduces the resource utilization in cilium agent.
+
+#### Cons
+
+* Higher latency in DNS resolution as the DNS queries are now handled by a separate process.
+
 ### Q: Restoring L7 DNS policy when agent is unavailable
 
-In an event where SDP restarts when agent's gRPC service is unavailable, can / should SDP rely on information stored on the filesystem to restore per endpoint L7 policy ?
+In an event where SDP restarts when agent's gRPC service is unavailable. SDP can read from the filesystem and restore the policy. The state is stored in `ep_config.json` file. So the flow would be first trying to connect to agent and if it fails, read from the file.
```

Comment:
What about the case where the pod hasn't initiated a connection yet ? We noticed that the JSON and the header files are updated only lazily right ?

Reply:
Yeah I think reading from file is a fallback if agent is not available and what we have in the file is the best we can do until we start writing to it.